### PR TITLE
Task: add links for landing page directory

### PR DESCRIPTION
**Pull request by oubiwann**
This branch is the culmination of the work to support a directory of landing pages in the CMR search app that can be utilized by web crawlers for indexing search data. Several other task branches have already been merged; together, these provide the full feature.

### CODE REVIEW

File: system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+      (is (= 200 (:status response)))
+      (is (substring? expected-header-link (:body response))))))
+
+(deftest top-level-links
+  (testing "check top level links"
+    (let [url (str base-url "/site/collections/directory")
+          response (client/get url)
+          body (:body response)]
+      (is (= 200 (:status response)))
+      ;; The collections not taged with eosdis shouldn't show up
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: tagged typo

File: system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+            [cmr.system-int-test.utils.index-util :as index]
+            [cmr.system-int-test.utils.ingest-util :as ingest]
+            [cmr.system-int-test.utils.tag-util :as tags]
+            [cmr.mock-echo.client.echo-util :as e]
+            [cmr.system-int-test.system :as s]
+            [cmr.transmit.config :as transmit-config]
+            [cmr.umm-spec.models.umm-common-models :as cm]
+            [cmr.umm-spec.test.expected-conversion :as exp-conv]))
+
+(def ^:private base-url (format "%s://%s:%s"
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: I think the base-url should end with a "/" to match what is being used in the source code though I'm guessing all of the tests that check this base-url are probably adding a "/" when used.

File: system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+      ;; The collections not taged with eosdis shouldn't show up
+      (is (not (substring? notexpected-provider-level-link body)))
+      ;; This page should also have a header link
+      (is (substring? expected-header-link body)))))
+
+;; Note that the following test was originally in the unit tests for this code
+;; (thus the similarity of it to those tests) but had to be moved to an
+;; integration test with the introduction of `base-url` support in the
+;; templates (which the following text exercises). The base URL is obtained
+;; (ulimately) by calling c.t.config/application-public-root-url which needs
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: ultimately

File: system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+;; `public-conf` data set in both the route-creation as well as the request. It
+;; was thus just easier and more natural to perform the required test as part
+;; the integration tests, since the running system already has that data set
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: part of

File: search-app/src/cmr/search/site/data.clj [hunk 5]
```diff
@@ -43,15 +115,36 @@
...
+      (base-page context)
+      {:providers (map (partial provider-data "gov.nasa.eosdis") providers)})))
+
+(defn get-provider-tag-landing-links
+  "Generate the data necessary to render EOSDIS landing page links."
+  [context provider-id tag]
+  (let [query (query-svc/make-concepts-query
+               context
+               :collection
+               {:tag_key tag
```
Comment (oubiwann) [inline search-app/src/cmr/search/site/data.clj]: Oh, nice -- I actually much prefer `:tag-key` ... it's more Lispy ;-)

File: system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj (hunk 4 above)
Comment (oubiwann) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: Fixed.

File: system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj (hunk 3 above)
Comment (oubiwann) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: Fixed.

File: system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj (hunk 1 above)
Comment (oubiwann) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: Fixed (5 instances of this typo).

File: system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+            [cmr.umm-spec.test.expected-conversion :as exp-conv]))
+
+(def ^:private base-url (format "%s://%s:%s"
+                                (transmit-config/search-protocol)
+                                (transmit-config/search-host)
+                                (transmit-config/search-port)))
+
+;;; General utility functions for the tests
+
+(defn- substring?
```
Comment (oubiwann) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: Oh, sweet ... about time!

Comment (oubiwann) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: Fixed.

File: system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj (hunk 2 above)
Comment (oubiwann) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: Good point. Fixed.

Comment (oubiwann) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: Done. (To not keep you in suspense: `application-public-root-url` requires a context; we're just using the HTTP client for these tests, so no context.)

File: system-int-test/src/cmr/system_int_test/utils/ingest_util.clj [hunk 7]
```diff
@@ -600,29 +600,39 @@
   {:grant-all-search? true
    :grant-all-ingest? true})
 
-(defn reset-fixture
-  "Resets all the CMR systems then creates the given providers in ECHO and the CMR.
-  providers can be passed in two ways: 1) a map of provider-guids to provider-ids
-  {'provider-guid1' 'PROV1' 'provider-guid2' 'PROV2'}, or
+(defn setup-providers
+  "Creates the given providers in ECHO and the CMR. Providers can be passed in
```
Comment (oubiwann) [inline system-int-test/src/cmr/system_int_test/utils/ingest_util.clj]: Fixed.

File: search-app/test/cmr/search/test/site/data.clj [hunk 8]
```diff
@@ -14,3 +57,97 @@
           concept-id "C1200196931-SCIOPS"]
       (is (= (data/cmr-link cmr-host concept-id)
              "https://cmr.sit.earthdata.nasa.gov/concepts/C1200196931-SCIOPS.html")))))
+
+(deftest get-doi
+  (testing "try to get the doi entry from data that doesn't have one"
+    (is (not (data/get-doi coll-data-1))))
+  (testing "get the doi entry"
+    (is (= (data/get-doi coll-data-2)
```
Comment (oubiwann) [inline search-app/test/cmr/search/test/site/data.clj]: Yup, you're absolutely correct. Fixed.

File: search-app/test/cmr/search/test/site/data.clj [hunk 9]
```diff
@@ -1,6 +1,49 @@
...
+    {:provider-id "PROV1"
+     :concept-id "C1200000003-PROV1"}
+   :umm
+    {"ShortName" "s3"
+     "EntryTitle" "coll3"
+     "DOI"
+     {"Authority" "auth6"
+      "DOI" "doi6"}}})
+
+(def coll-data-3
```
Comment (oubiwann) [inline search-app/test/cmr/search/test/site/data.clj]: Removed.

File: search-app/src/cmr/search/site/pages.clj [hunk 10]
```diff
@@ -50,3 +50,12 @@
    context
    "templates/eosdis-directory-links.html"
    (data/get-eosdis-directory-links context)))
+
+(defn proivider-tag-directory
```
Comment (oubiwann) [inline search-app/src/cmr/search/site/pages.clj]: Fixed (also fixed the reference in routes).

File: search-app/src/cmr/search/site/data.clj [hunk 11]
```diff
@@ -25,8 +57,48 @@
...
+(defn get-long-name
+  "Get a collection item's long name."
+  [item]
+  (or (get-in item [:umm "EntryTitle"])
+      (get-in item [:meta :concept-id])))
+
+(defn get-short-name
+  "Get a collection item's short name, if it exists."
+  [item]
+  (if-let [short-name (get-in item [:umm "ShortName"])]
```
Comment (oubiwann) [inline search-app/src/cmr/search/site/data.clj]: Removed.

File: search-app/src/cmr/search/site/data.clj (hunk 5 above)
Comment (oubiwann) [inline search-app/src/cmr/search/site/data.clj]: Fixed.

File: search-app/src/cmr/search/site/data.clj [hunk 12]
```diff
@@ -25,8 +57,48 @@
...
+  "Create the `href` part of a landing page link."
+  [cmr-base-url item]
+  (if (has-doi? item)
+    (doi-link (get-doi item))
+    (cmr-link cmr-base-url (get-in item [:meta :concept-id]))))
+
+(defn get-long-name
+  "Get a collection item's long name."
+  [item]
+  (or (get-in item [:umm "EntryTitle"])
```
Comment (oubiwann) [inline search-app/src/cmr/search/site/data.clj]: Fixed. Also renamed `long-name` instances to `entity-title`, per your suggestion in chat.

File: search-app/src/cmr/search/site/data.clj [hunk 13]
```diff
@@ -10,12 +10,44 @@
...
+
+(defn get-doi
+  "Extract the DOI information from a collection item."
+  [item]
+  (get-in item [:umm "DOI"]))
+
+(defn has-doi?
+  "Determine wheter a collection item has a DOI entry."
+  [item]
+  (if-not (nil? (get-doi item))
```
Comment (oubiwann) [inline search-app/src/cmr/search/site/data.clj]: Nice catch. Changed.

File: search-app/src/cmr/search/site/data.clj [hunk 14]
```diff
@@ -10,12 +10,44 @@
...
+   :name (:provider-id provider-data)
+   :tag tag})
+
+(defn get-doi
+  "Extract the DOI information from a collection item."
+  [item]
+  (get-in item [:umm "DOI"]))
+
+(defn has-doi?
+  "Determine wheter a collection item has a DOI entry."
```
Comment (oubiwann) [inline search-app/src/cmr/search/site/data.clj]: Fixed.

**Comment by oubiwann** [comment]
Addressed all feedback, pushed, and waiting for Bamboo to go green ...

File: system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj [hunk 15]
```diff
@@ -0,0 +1,242 @@
+(ns cmr.system-int-test.search.collection-directory-pages-test
+  "This tests a running CMR site's directory links at all levels: top-most,
+  eosdis, and provider."
+  (:require [clj-http.client :as client]
+            [clojure.test :refer :all]
+            [clojure.string :as string]
```
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: nit: CMR naming convention for this is `[clojure.string :as str]`.

File: search-app/src/cmr/search/services/query_service.clj [hunk 16]
```diff
@@ -106,35 +106,51 @@
...
+  (or (:tag-data params) (:tag_data params)))
+
+(defn make-concepts-query
+  "Utility function for generating an elastic-ready query."
+  ([context concept-type params]
+    (->> params
+         (make-concepts-tag-data)
+         (make-concepts-query
+           context concept-type params)))
+  ([context concept-type params tag-data]
```
Comment (indiejames) [inline search-app/src/cmr/search/services/query_service.clj]: This use of multi-arity seems a little unconventional. Would anyone ever call this function and pass in the tag-data parameter? Usually I see this used for optional parameters (like an options map).

File: system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj (hunk 15 above)
Comment (oubiwann) [inline system-int-test/test/cmr/system_int_test/search/collection_directory_pages_test.clj]: After a quick chat, it seems like we're moving towards more standard require-as aliases ...

File: search-app/src/cmr/search/services/query_service.clj (hunk 16 above)
Comment (oubiwann) [inline search-app/src/cmr/search/services/query_service.clj]: Yeah, in fact the multi-arity here was added to support the original usage of this code (passing the `tag-data` parameter).

I've seen multi-arity functions used a ton for additional args. In fact, one of the most common patterns I've seen for Clojure multi-arity functions is for lower-arity to pass defaults to higher-arity.

Comment (indiejames) [inline search-app/src/cmr/search/services/query_service.clj]: Yes, that's the most common usage I have seen as well. I didn't realize the original function here took a tag-data parameter.

Comment (oubiwann) [inline search-app/src/cmr/search/services/query_service.clj]: Sweet.

**Pull request merged**